From d69b5c461ede6a3caf31a6cec0201e2423ba0504 Mon Sep 17 00:00:00 2001
From: Alexander Couzens <lynxis@fe80.eu>
Date: Mon, 16 Mar 2015 02:49:48 +0100
Subject: [PATCH 2/9] target/omap/image: add sdcard image based on sunxi

Flash the sdcard image via `dd if=openwrt-omap-beagleboneblack-sdcard-vfat-am335x_evm.img of=/dev/<sdcard device>`
---
 target/linux/omap/image/Config.in              |  5 ++++
 target/linux/omap/image/Makefile               | 37 ++++++++++++++++++++++++++
 target/linux/omap/image/gen_omap_sdcard_img.sh | 33 +++++++++++++++++++++++
 target/linux/omap/image/uEnv.txt               |  6 +++++
 4 files changed, 81 insertions(+)
 create mode 100644 target/linux/omap/image/Config.in
 create mode 100755 target/linux/omap/image/gen_omap_sdcard_img.sh
 create mode 100644 target/linux/omap/image/uEnv.txt

diff --git a/target/linux/omap/image/Config.in b/target/linux/omap/image/Config.in
new file mode 100644
index 0000000..08b88eb
--- /dev/null
+++ b/target/linux/omap/image/Config.in
@@ -0,0 +1,5 @@
+config OMAP_SD_BOOT_PARTSIZE
+	int "Boot (SD Card) filesystem partition size (in MB)"
+	depends on TARGET_omap
+	default 20
+
diff --git a/target/linux/omap/image/Makefile b/target/linux/omap/image/Makefile
index ed3d90d..bca00eb 100644
--- a/target/linux/omap/image/Makefile
+++ b/target/linux/omap/image/Makefile
@@ -10,12 +10,48 @@ include $(INCLUDE_DIR)/image.mk
 UBIFS_OPTS = -F -m 2048 -e 124KiB -c 4096 -U
 UBI_OPTS = -m 2048 -p 128KiB -s 512 -O 2048
 
+FAT32_BLOCK_SIZE=1024
+FAT32_BLOCKS=$(shell echo $$(($(CONFIG_OMAP_SD_BOOT_PARTSIZE)*1024*1024/$(FAT32_BLOCK_SIZE))))
+
+# $1 - name of bootloader directory under BIN_DIR
+# $2 - output filename part
+define Image/Build/SDCard
+	rm -f $(KDIR)/boot-$(2).img
+	mkdosfs $(KDIR)/boot-$(2).img -C $(FAT32_BLOCKS)
+	mcopy -i $(KDIR)/boot-$(2).img $(BIN_DIR)/uboot-omap-$(1)/MLO ::MLO
+	mcopy -i $(KDIR)/boot-$(2).img $(BIN_DIR)/uboot-omap-$(1)/u-boot.img ::u-boot.img
+	mcopy -i $(KDIR)/boot-$(2).img -s $(BIN_DIR)/dtbs/ ::dtbs
+	mcopy -i $(KDIR)/boot-$(2).img $(BIN_DIR)/openwrt-omap-zImage ::zImage
+	mcopy -i $(KDIR)/boot-$(2).img $(BIN_DIR)/openwrt-omap-uImage ::uImage
+	mcopy -i $(KDIR)/boot-$(2).img ./uEnv.txt ::uEnv.txt
+	./gen_omap_sdcard_img.sh \
+		$(BIN_DIR)/$(IMG_PREFIX)-$(2)-sdcard-vfat-$(1).img \
+		$(KDIR)/boot-$(2).img \
+		$(KDIR)/root.ext4 \
+		$(CONFIG_OMAP_SD_BOOT_PARTSIZE) \
+		$(CONFIG_TARGET_ROOTFS_PARTSIZE)
+endef
+
 define Image/BuildKernel
 	$(CP) $(KDIR)/zImage $(BIN_DIR)/openwrt-$(BOARD)-zImage
  ifneq ($(CONFIG_TARGET_ROOTFS_INITRAMFS),)
 	$(CP) $(KDIR)/zImage-initramfs $(BIN_DIR)/openwrt-$(BOARD)-zImage-initramfs
  endif
 
+	mkimage -A arm -O linux -T kernel -C none \
+		-a 0x80008000 -e 0x80008000 \
+		-n 'ARM OpenWrt Linux-$(LINUX_VERSION)' \
+		-d $(KDIR)/zImage $(BIN_DIR)/$(IMG_PREFIX)-uImage
+
+    ifneq ($(CONFIG_TARGET_ROOTFS_INITRAMFS),)
+	$(CP) $(KDIR)/zImage-initramfs $(BIN_DIR)/$(IMG_PREFIX)-zImage-initramfs
+	echo -ne '\x00\x00\x00\x00' >> $(BIN_DIR)/$(IMG_PREFIX)-zImage-initramfs
+	$(call Image/BuildKernel/MkuImage, \
+		none, 0x80008000, 0x80008000, \
+		$(BIN_DIR)/$(IMG_PREFIX)-zImage-initramfs, \
+		$(BIN_DIR)/$(IMG_PREFIX)-uImage-initramfs \
+	)
+    endif
  ifneq ($(CONFIG_TARGET_ROOTFS_INCLUDE_KERNEL),)
 	$(INSTALL_DIR) $(TARGET_DIR)/boot
 	$(CP) $(BIN_DIR)/openwrt-$(BOARD)-zImage $(TARGET_DIR)/boot/zImage
@@ -34,6 +70,7 @@ endef
 
 define Image/Build
 	$(call Image/Build/$(1),$(1))
+	$(call Image/Build/SDCard,am335x_evm,beagleboneblack)
 endef
 
 define Image/Build/jffs2-64k
diff --git a/target/linux/omap/image/gen_omap_sdcard_img.sh b/target/linux/omap/image/gen_omap_sdcard_img.sh
new file mode 100755
index 0000000..c2f2aad
--- /dev/null
+++ b/target/linux/omap/image/gen_omap_sdcard_img.sh
@@ -0,0 +1,33 @@
+#!/usr/bin/env bash
+
+#
+# Copyright (C) 2013 OpenWrt.org
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+
+set -x
+[ $# -eq 5 ] || {
+    echo "SYNTAX: $0 <file> <bootfs image> <rootfs image> <bootfs size> <rootfs size>"
+    exit 1
+}
+
+OUTPUT="$1"
+BOOTFS="$2"
+ROOTFS="$3"
+BOOTFSSIZE="$4"
+ROOTFSSIZE="$5"
+
+head=4
+sect=63
+
+set `ptgen -o $OUTPUT -h $head -s $sect -l 1024 -t c -p ${BOOTFSSIZE}M -t 83 -p ${ROOTFSSIZE}M`
+
+BOOTOFFSET="$(($1 / 512))"
+BOOTSIZE="$(($2 / 512))"
+ROOTFSOFFSET="$(($3 / 512))"
+ROOTFSSIZE="$(($4 / 512))"
+
+dd bs=512 if="$BOOTFS" of="$OUTPUT" seek="$BOOTOFFSET" conv=notrunc
+dd bs=512 if="$ROOTFS" of="$OUTPUT" seek="$ROOTFSOFFSET" conv=notrunc
diff --git a/target/linux/omap/image/uEnv.txt b/target/linux/omap/image/uEnv.txt
new file mode 100644
index 0000000..976f77a
--- /dev/null
+++ b/target/linux/omap/image/uEnv.txt
@@ -0,0 +1,6 @@
+bootpart=0:1
+bootdir=/
+fdtdir=/dtbs
+uenvcmd=run loadfdt; run loadimage; run mmcargs;  bootz ${loadaddr} - ${fdtaddr}
+
+loadfdt=load mmc ${bootpart} ${fdtaddr} ${fdtdir}/${fdtfile}
-- 
2.3.3

